lib/src/working_copy.rs

```diff
@@ -132,6 +134,20 @@ fn file_states_from_proto(
     file_states
 }
 
+fn sparse_patterns_from_proto(proto: &crate::protos::working_copy::TreeState) -> Vec<RepoPath> {
+    let mut sparse_patterns = vec![];
+    if proto.has_sparse_patterns() {
```

Comment:
protobufs, if that's what these are, have a difference between "has foo" and "foo is non-empty". Should this be:

```
if proto.has_sparse_patterns() {
  ...
}

// For compatibility with old working copies, and empty sparse patterns.
if sparse_patterns.is_empty() {
  sparse_patterns.push(RepoPath::root());
}

sparse_patterns
```

Reply:
No, the idea is that the presence of the `sparse_pattern` field indicates that the working copy was written by a version after this commit. The goal is to allow and empty list of patterns.